cli/tools/bundle/mod.rs

```diff
@@ -0,0 +1,879 @@
+use deno_npm::npm_rc::ResolvedNpmRc;
+use deno_npm::registry::NpmRegistryApi;
+use deno_npm_cache::TarballCache;
+use deno_resolver::npm::managed::ResolvePkgFolderFromDenoModuleError;
+use deno_resolver::workspace::WorkspaceNpmPatchPackages;
+use deno_runtime::deno_permissions::PermissionsContainer;
+use deno_semver::npm::NpmPackageReqReference;
+use deno_semver::package::PackageNv;
+use esbuild_rs::protocol;
+use esbuild_rs::EsbuildFlagsBuilder;
+use esbuild_rs::EsbuildService;
+use indexmap::IndexMap;
+use node_resolver::errors::PackageSubpathResolveError;
+use node_resolver::NodeResolutionKind;
+use node_resolver::ResolutionMode;
+use sys_traits::EnvCurrentDir;
+
+use crate::args::BundleFlags;
+use crate::args::BundleFormat;
+use crate::args::Flags;
+use crate::args::PackageHandling;
+use crate::cache::DenoDir;
+use crate::factory::CliFactory;
+use crate::graph_container::MainModuleGraphContainer;
+use crate::graph_container::ModuleGraphContainer;
+use crate::graph_container::ModuleGraphUpdatePermit;
+use crate::module_loader::ModuleLoadPreparer;
+use crate::module_loader::PrepareModuleLoadOptions;
+use crate::node::CliNodeResolver;
+use crate::npm::CliNpmCache;
+use crate::npm::CliNpmCacheHttpClient;
+use crate::npm::CliNpmRegistryInfoProvider;
+use crate::npm::CliNpmResolver;
+use crate::resolver::CliNpmReqResolver;
+use crate::resolver::CliResolver;
+use crate::sys::CliSys;
+
+const ESBUILD_VERSION: &str = "0.25.5";
+
+pub async fn ensure_esbuild(
+  deno_dir: &DenoDir,
+  npmrc: &ResolvedNpmRc,
+  npm_registry_info: &Arc<CliNpmRegistryInfoProvider>,
+  workspace_patch_packages: &Arc<WorkspaceNpmPatchPackages>,
+  tarball_cache: &Arc<TarballCache<CliNpmCacheHttpClient, CliSys>>,
+  npm_cache: &CliNpmCache,
+  target: &str,
+) -> Result<PathBuf, AnyError> {
+  let esbuild_path = deno_dir
+    .dl_folder_path()
+    .join(format!("esbuild-{}", ESBUILD_VERSION))
+    .join(format!("esbuild-{}", target));
+
+  if esbuild_path.exists() {
+    return Ok(esbuild_path);
+  }
+  let pkg_name = format!("@esbuild/{}", target);
+  let nv =
+    PackageNv::from_str(&format!("{}@{}", pkg_name, ESBUILD_VERSION)).unwrap();
+  let api = npm_registry_info.as_npm_registry_api();
+  let info = api.package_info(&pkg_name).await?;
+  let version_info = info.version_info(&nv, &workspace_patch_packages.0)?;
+  if let Some(dist) = &version_info.dist {
+    let registry_url = npmrc.get_registry_url(&nv.name);
+    let package_folder =
+      npm_cache.package_folder_for_nv_and_url(&nv, registry_url);
+    let existed = package_folder.exists();
+
+    if !existed {
+      tarball_cache.ensure_package(&nv, &dist).await?;
+    }
+
+    let path = package_folder.join("bin/esbuild");
+
+    std::fs::create_dir_all(esbuild_path.parent().unwrap())?;
+    std::fs::copy(&path, &esbuild_path)?;
+
+    if !existed {
+      std::fs::remove_dir_all(&package_folder)?;
+    }
+    return Ok(esbuild_path);
+  }
+  anyhow::bail!("esbuild not found");
```

Comment:
Need a better error here - preferably with a suggestion for a solution how to unblock oneself